src/semafour.cc

```diff
@@ -91,6 +96,27 @@ int Semafour::Wait() {
 }
 
 
+int Semafour::TryWait() {
+  int r;
+
+  do {
+    r = sem_trywait(_sem);
+  } while (r == -1 && errno == EINTR);
+
+  bool locked = r == -1 && errno == EAGAIN;
```

Comment:
This logic doesn't seem correct to me. Couldn't you just return `0` or `-errno` like `Semafour::Wait()` does? That would also mean the `TRYWAIT_OK` and `TRYWAIT_LOCKED` constants wouldn't be needed.

Reply:
There are three states here:
1 - The lock was acquired: currently, `kTryWaitOk` is returned
2 - The lock wasn't acquired: currently, `kTryWaitLocked` is returned
3 - An unexpected error occured: currently, `-errno` is returned

I wanted the function doing the dirty work to actually reflect that, because I think it is more declarative. Returning only `0` or `errno` would mean pushing the responsibility of decoding the messy response code in the consumer functions, which I think is not their purpose. Should we write another function that uses `tryWait`, we won't have to duplicate the "business logic" of checking for `EAGAIN`.

My approach here was to try to copy `wait()`, just extending it to have two possible valid states instead of `wait()`'s single OK return code.

Ideally, we should use something like [the Either construct](https://hackage.haskell.org/package/category-extras-0.52.0/docs/Control-Monad-Either.html), but I'd say introducing it in such a small project wouldn't actually make any sense.

Now that I've tried to make my case: would you still prefer it if I did exactly like in `wait()`, and move the `EAGAIN` check to the consumer function, or to a dedicated (inline) function?

Reply:
The meaning of the `locked` boolean is definitely ambiguous though. What this means is: `was the semaphore already locked when we tried to acquire it?`

Reply:
I will.